### PR TITLE
Remove unnecessary names in Simpy Grammar

### PR DESCRIPTION
**Pull request by lysnikolaou**
Due to #121, we can remove lots of unnecessary names in the Grammar. Merging unnecessary alternatives will come in another PR.

### CODE REVIEW

File: data/simpy.gram [hunk 1]
```diff
@@ -114,90 +110,90 @@ dotted_name[expr_ty]:
...
-    | a='if' b=named_expression ':' c=block e=elif_stmt {
-        _Py_If(b, c, singleton_seq(p, e), EXTRA) }
-    | a='if' b=named_expression ':' c=block {
-        _Py_If(b, c, NULL, EXTRA) }
+    | 'if' a=named_expression ':' b=block c=else_block {
+        _Py_If(a, b, c, EXTRA) }
+    | 'if' a=named_expression ':' b=block c=elif_stmt {
+        _Py_If(a, b, singleton_seq(p, c), EXTRA) }
+    | 'if' a=named_expression ':' b=block {
+        _Py_If(a, b, NULL, EXTRA) }
```
Comment (gvanrossum) [inline data/simpy.gram]: The first and 3rd alternatives (if-else and if without anything) can be combined into one using `c=[else_block]`. (The elif variant should go first to deal with the ambiguity.)

File: data/simpy.gram [hunk 2]
```diff
@@ -114,90 +110,90 @@ dotted_name[expr_ty]:
...
-    | 'elif' b=named_expression ':' c=block e=elif_stmt {
-        _Py_If(b, c, singleton_seq(p, e), EXTRA) }
-    | 'elif' b=named_expression ':' c=block {
-        _Py_If(b, c, NULL, EXTRA) }
+    | 'elif' a=named_expression ':' b=block c=else_block {
+        _Py_If(a, b, c, EXTRA) }
+    | 'elif' a=named_expression ':' b=block c=elif_stmt {
+        _Py_If(a, b, singleton_seq(p, c), EXTRA) }
+    | 'elif' a=named_expression ':' b=block {
+        _Py_If(a, b, NULL, EXTRA) }
```
Comment (gvanrossum) [inline data/simpy.gram]: Ditto.

File: data/simpy.gram [hunk 3]
```diff
@@ -114,90 +110,90 @@ dotted_name[expr_ty]:
...
 
 while_stmt[stmt_ty]:
-    | a='while' ex=named_expression ':' b=block el=else_block {
-        _Py_While(ex, b, el, EXTRA) }
-    | a='while' ex=named_expression ':' b=block {
-        _Py_While(ex, b, NULL, EXTRA) }
+    | 'while' a=named_expression ':' b=block c=else_block {
+        _Py_While(a, b, c, EXTRA) }
+    | 'while' a=named_expression ':' b=block {
+        _Py_While(a, b, NULL, EXTRA) }
```
Comment (gvanrossum) [inline data/simpy.gram]: These can also be combined.

File: data/simpy.gram [hunk 4]
```diff
@@ -114,90 +110,90 @@ dotted_name[expr_ty]:
...
+    | ASYNC 'with' a=','.with_item+ ':' b=block {
+        _Py_AsyncWith(a, b, NULL, EXTRA) }
+    | 'with' '(' a=','.with_item+ ')' ':' b=block {
+        _Py_With(a, b, NULL, EXTRA) }
+    | 'with' a=','.with_item+ ':' b=block {
+        _Py_With(a, b, NULL, EXTRA) }
 with_item[withitem_ty]:
     | e=expression o=['as' t=target { t }] { _Py_withitem(e, store_name(p, o), p->arena) }
 
 try_stmt[stmt_ty]:
```
Comment (gvanrossum) [inline data/simpy.gram]: I suspect that some (but not all) try-except alternatives can also be combined.

File: data/simpy.gram [hunk 5]
```diff
@@ -114,90 +110,90 @@ dotted_name[expr_ty]:
...
+    | 'return' a=expressions { _Py_Return(a, EXTRA) }
+    | 'return' { _Py_Return(NULL, EXTRA) }
 
 raise_stmt[stmt_ty]:
-    | a='raise' b=expression 'from' c=expression { _Py_Raise(b, c, EXTRA) }
-    | a='raise' b=expression { _Py_Raise(b, NULL, EXTRA) }
-    | a='raise' { _Py_Raise(NULL, NULL, EXTRA) }
+    | 'raise' a=expression 'from' b=expression { _Py_Raise(a, b, EXTRA) }
+    | 'raise' a=expression { _Py_Raise(a, NULL, EXTRA) }
+    | 'raise' { _Py_Raise(NULL, NULL, EXTRA) }
```
Comment (gvanrossum) [inline data/simpy.gram]: This ought to be combinable into two:
```
    | 'raise' a=expression 'from' c=expression {...}
    | 'raise' a=[expression] {...}
```
(Alas, I can't figure out how to make it a single one. I filed #129 for a possible solution, but i think that would require significant refactoring of how the generated code for nested groups looks.)

File: data/simpy.gram [hunk 6]
```diff
@@ -114,90 +110,90 @@ dotted_name[expr_ty]:
...
         _Py_AsyncFunctionDef(((expr_ty) n)->v.Name.id, params, b, NULL, a, NULL, EXTRA) }
-    | s=ASYNC 'def' n=NAME '(' ')' a=['->' z=annotation { z }] ':' b=block {
+    | ASYNC 'def' n=NAME '(' ')' a=['->' z=annotation { z }] ':' b=block {
         _Py_AsyncFunctionDef(((expr_ty) n)->v.Name.id, empty_arguments(p), b, NULL, a, NULL, EXTRA) }
-    | s='def' n=NAME '(' params=parameters ')' a=['->' z=annotation { z }] ':' b=block {
+    | 'def' n=NAME '(' params=parameters ')' a=['->' z=annotation { z }] ':' b=block {
         _Py_FunctionDef(((expr_ty) n)->v.Name.id, params, b, NULL, a, NULL, EXTRA) }
-    | s='def' n=NAME '(' ')' a=['->' z=annotation { z }] ':' b=block {
+    | 'def' n=NAME '(' ')' a=['->' z=annotation { z }] ':' b=block {
         _Py_FunctionDef(((expr_ty) n)->v.Name.id, empty_arguments(p), b, NULL, a, NULL, EXTRA) }
```
Comment (gvanrossum) [inline data/simpy.gram]: I think these could be collapsed with judicious use of `params ? params : empty_arguments(p)`. Ditto for the ASYNC version above.

File: data/simpy.gram [hunk 7]
```diff
@@ -261,8 +257,8 @@ expression[expr_ty]:
     | disjunction
 
 lambdef[expr_ty]:
-    | a='lambda' ':' c=expression { _Py_Lambda(empty_arguments(p), c, EXTRA) }
-    | a='lambda' b=lambda_parameters ':' c=expression { _Py_Lambda(b, c, EXTRA) }
+    | 'lambda' ':' a=expression { _Py_Lambda(empty_arguments(p), a, EXTRA) }
+    | 'lambda' a=lambda_parameters ':' b=expression { _Py_Lambda(a, b, EXTRA) }
```
Comment (gvanrossum) [inline data/simpy.gram]: Same?

File: data/simpy.gram [hunk 8]
```diff
@@ -412,9 +408,9 @@ for_if_clauses[asdl_seq*]:
         { _Py_comprehension(a, b, c, (y == NULL) ? 0 : 1, p->arena) })+ { a }
 
 yield_expr[expr_ty]:
-    | a='yield' 'from' b=expression { _Py_YieldFrom(b, EXTRA) }
-    | a='yield' b=expressions { _Py_Yield(b, EXTRA) }
-    | a='yield' { _Py_Yield(NULL, EXTRA) }
+    | 'yield' 'from' a=expression { _Py_YieldFrom(a, EXTRA) }
+    | 'yield' a=expressions { _Py_Yield(a, EXTRA) }
+    | 'yield' { _Py_Yield(NULL, EXTRA) }
```
Comment (gvanrossum) [inline data/simpy.gram]: The last two could be collapsed.

**Comment by lysnikolaou** [comment]
@gvanrossum, I was thinking of doing that in another PR, should I do it here instead?

**Comment by gvanrossum** [comment]
That’s fine too.
-- 
--Guido (mobile)

**Pull request merged**